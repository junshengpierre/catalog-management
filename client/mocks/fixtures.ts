export const productList = [
  {
    image:
      'http://localhost:5000/images/sample-image49a417fe-e018-4c93-a417-52bb84fa6917.png',
    title: 'Title 1',
    description: 'Description 1',
    quantity: 100,
    price: 100,
    status: 'Public',
    createdAt: '2020-11-09T14:03:37.152Z',
    updatedAt: '2020-11-09T14:03:37.152Z',
    id: '5fa94c39c4d4033a35699c99',
  },
  {
    image:
      'http://localhost:5000/images/sample-image54f3b5b8-9389-4ef9-a360-8fd7fc5602d0.png',
    title: 'Title 2',
    description: 'Description 2',
    quantity: 200,
    price: 200,
    status: 'Private',
    createdAt: '2020-11-09T14:54:25.580Z',
    updatedAt: '2020-11-09T14:54:25.580Z',
    id: '5fa95821c4d4033a35699c9a',
  },
]
