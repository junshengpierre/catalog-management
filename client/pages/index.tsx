import {
  Button,
  Container,
  Card,
  Alert,
  CardColumns,
  Spinner,
  Modal,
} from 'react-bootstrap'
import { MainLayout } from '../components/MainLayout'
import styled from '@emotion/styled'
import { useQuery } from 'react-query'
import isEmpty from 'lodash/isEmpty'
import { Product } from '../types'
import Link from 'next/link'
import { formatDisplayPrice } from '../utils/formatDisplayPrice'
import { api } from '../api'
import { useState, useRef } from 'react'
import { AddProductForm } from '../components/AddProductForm'

export const Home = (): JSX.Element => {
  const { data, isLoading, isError } = useQuery('productList', async () => {
    const { data } = await api.get('/product')
    return data
  })

  const [showAddModal, setShowAddModal] = useState(false)

  const handleAddModalClose = () => setShowAddModal(false)
  const handleAddModalShow = () => setShowAddModal(true)

  return (
    <MainLayout>
      <Container>
        {isError && !data && (
          <Alert className="mt-4" variant="danger">
            Oops! Something went wrong.
          </Alert>
        )}

        <Container className="mt-4 px-0 d-flex justify-content-between align-items-center">
          <Heading>Products</Heading>
          <Button onClick={handleAddModalShow} data-testid="addProductButton">
            Add Product
          </Button>
        </Container>

        <Container className="px-0 mt-4 mb-5">
          {isLoading && !data && (
            <Container className="px-0 py-4 d-flex justify-content-center">
              <Spinner animation="border" role="status">
                <span className="sr-only">Loading...</span>
              </Spinner>
            </Container>
          )}

          {!isLoading && isEmpty(data) && (
            <div className="my-4 d-flex justify-content-center align-items-center">
              No Product Listings
            </div>
          )}

          {/* TODO: Fix broken grid display when number of items % 3 is 1 */}
          {Boolean(data) && (
            <CardColumns>
              {data.map((product: Product) => {
                return (
                  <Card key={product.id} data-testid="productListItem">
                    <Card.Img variant="top" src={product.image} />
                    <Card.Body>
                      <Card.Title>{product.title}</Card.Title>
                      <Card.Text>{product.description}</Card.Text>
                      <Card.Text>{`Quantity: ${product.quantity}`}</Card.Text>
                      <Container className="px-0 d-flex justify-content-between">
                        <Card.Text>{`Price: ${formatDisplayPrice(
                          product.price
                        )}`}</Card.Text>
                        <Card.Text>{`Status: ${product.status}`}</Card.Text>
                      </Container>
                    </Card.Body>
                    <Card.Footer className="d-flex justify-content-end">
                      <Link href={`/product/${product.id}`}>
                        <a>View Details</a>
                      </Link>
                    </Card.Footer>
                  </Card>
                )
              })}
            </CardColumns>
          )}
        </Container>
      </Container>
      <AddModal show={showAddModal} onClose={handleAddModalClose} />
    </MainLayout>
  )
}

export default Home

const Heading = styled.h1`
  font-size: 18px;
`

// lastName: Yup.string().required(),
// username: Yup.string().required(),
// city: Yup.string().required(),
// state: Yup.string().required(),
// zip: Yup.string().required(),
// terms: Yup.bool().required(),

const AddModal = ({
  show,
  onClose,
}: {
  show: boolean
  onClose: () => void
}) => {
  const formRef = useRef()

  return (
    <Modal size="lg" show={show} onHide={onClose}>
      <Modal.Header closeButton>
        <Modal.Title>Add Product</Modal.Title>
      </Modal.Header>

      <Modal.Body>
        <AddProductForm ref={formRef} onSuccess={onClose} />
      </Modal.Body>

      <Modal.Footer>
        <Button variant="secondary" onClick={onClose}>
          Cancel
        </Button>
        <Button
          form="addProductForm"
          variant="primary"
          type="submit"
          onClick={() => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
            // @ts-ignore
            formRef?.current?.submitForm()
          }}
        >
          Submit
        </Button>
      </Modal.Footer>
    </Modal>
  )
}
