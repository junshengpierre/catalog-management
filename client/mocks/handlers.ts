import { productList } from './fixtures'
import { rest } from 'msw'

// INFO: Unsure why `http://localhost` domain works.
// Maybe because process.env.API_URI is undefined
// in test environment and we might need to find a
// way to mock it.
export const handlers = [
  rest.get('http://localhost/product', (req, res, ctx) => {
    return res(ctx.json(productList))
  }),
]
