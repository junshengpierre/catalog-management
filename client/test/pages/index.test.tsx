import React from 'react'
import { render, waitFor } from '../testUtils'
import { Home } from '../../pages/index'
import { productList } from '../../mocks/fixtures'

describe('Home page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Home />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  // TOOD: Implement failure handling tests
  // REf: https://github.com/mswjs/msw/issues/104#issuecomment-635050088
  it('clicking button triggers alert', async () => {
    const { getByText, queryByText, queryAllByText, getAllByTestId } = render(
      <Home />,
      {}
    )

    expect(getByText('Loading...')).toBeDefined()
    expect(getByText('Catalog Management System')).toBeDefined()

    await waitFor(() =>
      expect(getAllByTestId('productListItem').length).toEqual(2)
    )

    expect(queryByText('Loading...')).toBeNull()
    expect(queryByText(productList[0].title)).not.toBeNull()
    expect(queryByText(productList[0].description)).not.toBeNull()
    expect(queryAllByText(/Quantity/)).not.toBeNull()
    expect(queryAllByText(/Price/)).not.toBeNull()
    expect(queryAllByText(/Status/)).not.toBeNull()
  })
})
