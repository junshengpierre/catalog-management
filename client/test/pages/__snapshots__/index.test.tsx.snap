// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column h-100"
  >
    <main
      class="css-1122yjz"
    >
      <div
        class="bg-light px-0 container-fluid"
      >
        <div
          class="container"
        >
          <nav
            class="px-0 navbar navbar-expand-lg navbar-light bg-transparent"
          >
            <a
              class="navbar-brand"
              href="/"
            >
              Catalog Management System
            </a>
          </nav>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="mt-4 px-0 d-flex justify-content-between align-items-center container"
        >
          <h1
            class="css-lc7fha"
          >
            Products
          </h1>
          <button
            class="btn btn-primary"
            data-testid="addProductButton"
            type="button"
          >
            Add Product
          </button>
        </div>
        <div
          class="px-0 mt-4 mb-5 container"
        >
          <div
            class="px-0 py-4 d-flex justify-content-center container"
          >
            <div
              class="spinner-border"
              role="status"
            >
              <span
                class="sr-only"
              >
                Loading...
              </span>
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="bg-dark text-white px-0 css-1e46tsl container-fluid"
    >
      <div
        class="py-3 d-flex justify-content-center container"
      >
        2020 Catalog Management System
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
